traffic_ops/app/db/migrations/2022011314075900_add_server_profile.up.sql

```diff
@@ -0,0 +1,29 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with this
+ * work for additional information regarding copyright ownership.  The ASF
+ * licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+CREATE TABLE IF NOT EXISTS public.server_profile (
+                                                     server bigint NOT NULL,
+                                                     profile_names text[] NOT NULL,
+                                                     priority bigint[] NOT NULL,
+    CONSTRAINT pk_server_profile PRIMARY KEY(profile_names, server),
+    CONSTRAINT fk_server_id FOREIGN KEY (server) REFERENCES public.server(id) ON DELETE CASCADE ON UPDATE CASCADE
+);
```

Comment:
Instead of
```
               Table "public.server_profile"
    Column     |   Type   | Collation | Nullable | Default 
---------------+----------+-----------+----------+---------
 server        | bigint   |           | not null | 
 profile_names | text[]   |           | not null | 
 priority      | bigint[] |           | not null | 
Indexes:
    "pk_server_profile" PRIMARY KEY, btree (profile_names, server)
Foreign-key constraints:
    "fk_server_id" FOREIGN KEY (server) REFERENCES server(id) ON UPDATE CASCADE ON DELETE CASCADE
```
Why not just 
```
           Table "public.server_profile"
  Column  |  Type  | Collation | Nullable | Default 
----------+--------+-----------+----------+---------
 server   | bigint |           | not null | 
 profile  | text   |           | not null | 
 priority | bigint |           | not null | 
Indexes:
    "pk_server_profile" PRIMARY KEY, btree (profile, server)
    "server_profile_unique_priority" UNIQUE CONSTRAINT, btree (server, priority)
Foreign-key constraints:
    "fk_server_profile_profile" FOREIGN KEY (profile) REFERENCES profile(name) ON UPDATE CASCADE ON DELETE RESTRICT
    "fk_server_profile_server" FOREIGN KEY (server) REFERENCES server(id) ON UPDATE CASCADE ON DELETE CASCADE
```
It's easier to work with, gives a slightly better guarantee of well-formed priorities (doesn't allow duplicates for the same server), and guarantees binding of Profile Names to actual Profiles - that is, under your current implementation, servers can have entries in their `profileNames` that aren't the names of any existing Profiles. This will happen if, for instance, the Name of a Profile is altered once it's been assigned to one or more Servers.

Reply:
You don't need to know what `btree` is to get that table, that's just an implementation detail (if you look above the current implementation's table also uses `btree` in its primary key, even though that term appears nowhere in your migration).

The priority and profile identity would still be arrays within the resulting Go structure, you'd just select them into an array from the multiple rows in the table. Even though you'd have to build an array in the read queries, I still think it's easier to use a table-based array because since the priority in the API is inherent in the list order, you have to use the priorities to sort the results which is much harder using two arrays instead of two columns.

Using your table, you select the arrays like this:
```sql
SELECT array_agg(profile_name)
FROM unnest(
	SELECT profile_names
	FROM public.server_profile
	WHERE "server" = $1
) WITH ORDINALITY AS profile_name
JOIN unnest(
	SELECT priority
	FROM public.server_profile
	WHERE "server" = $1
) WITH ORDINALITY AS "priority"
ON profile_name.ordinality = priority.ordinality
ORDER BY priority;
```

Using the table I suggested, you could just do:
```sql
SELECT array_agg(profile_name)
FROM public.server_profile
WHERE "server" = $1
ORDER BY "priority";
```

Reply:
I ran those through `explain` and apparently they're more or less equal in terms of performance. The latter is slightly lower cost, but only by like 1/14th of the total (and idk those units but that's probably insignificant).

To see these in action, I did a sample of each table type and each query to show they give the same output:

<details><summary><h2>Array columns</h2></summary>

### Table Data
```
 server |                 profile_names                 | priority 
--------+-----------------------------------------------+----------
      1 | {ATS_MID_XP,ATS_MID_LINEAR,ATS_MID_DS145HACK} | {1,2,3}
      2 | {A,B}                                         | {2,1}
(2 rows)
```

### Queries and Results

```sql
SELECT array_agg(profile_name ORDER BY priority.priority) AS profile_names
FROM sample_A,
unnest(sample_A.profile_names) WITH ORDINALITY AS profile_name
JOIN
unnest(sample_A.priority) WITH ORDINALITY AS "priority"
ON profile_name.ordinality = priority.ordinality
WHERE sample_A.server = 1;

                profile_names                   
-----------------------------------------------
 {ATS_MID_XP,ATS_MID_LINEAR,ATS_MID_DS145HACK}
(1 row)
```

```sql
SELECT array_agg(profile_name ORDER BY priority.priority) AS profile_names
FROM sample_A,
unnest(sample_A.profile_names) WITH ORDINALITY AS profile_name
JOIN
unnest(sample_A.priority) WITH ORDINALITY AS "priority"
ON profile_name.ordinality = priority.ordinality
WHERE sample_A.server = 2;

 profile_names 
---------------
 {B,A}
(1 row)
```

</details>

<hr/>

<details><summary><h2>Simple columns</h2></summary>

### Table Data

```
traffic_ops_development=# select * from sample_B;
 server |   profile_name    | priority 
--------+-------------------+----------
      1 | ATS_MID_XP        |        1
      1 | ATS_MID_LINEAR    |        2
      1 | ATS_MID_DS145HACK |        3
      2 | A                 |        2
      2 | B                 |        1
(5 rows)
```

### Queries and Results

```sql
SELECT array_agg(profile_name ORDER BY "priority") AS profile_names
FROM sample_b
WHERE "server" = 1;

                 profile_names                 
-----------------------------------------------
 {ATS_MID_XP,ATS_MID_LINEAR,ATS_MID_DS145HACK}
```

```sql
SELECT array_agg(profile_name ORDER BY "priority") AS profile_names
FROM sample_b
WHERE "server" = 2;

 profile_names 
---------------
 {B,A}
(1 row)
```

</details>

Reply:
I haven't been following this closely, but I believe @ocket8888 is correct -- we should generally avoid array-type columns (which are basically unique to PostgreSQL) in order to maintain referential integrity. This means we don't have referential integrity:
> under your current implementation, servers can have entries in their profileNames that aren't the names of any existing Profiles